Feature: Added support for Infomaniak kDrive

mafra99 opened the pull request:
**Resolved / Related Issues**

To prevent extra work, all changes to the Files codebase must link to an approved issue marked as `Ready to build`. Please insert the issue number following the hashtag with the issue number that this Pull Request resolves.
- Closes #15475

**Steps used to test these changes**

Stability is a top priority for Files and all changes are required to go through testing before being merged into the repo. Please include a list of steps that you used to test this PR.

1. kDrive is displayed in Files and copying of files etc. is possible


Note: The Sync Status column is not displayed, but this does not work in Windows File Explorer either. I think kDrive does not set the property [“System.FilePlaceholderStatus”](https://github.com/files-community/Files/blob/main/src/Files.App/ViewModels/ShellViewModel.cs#L1885) correctly, which Files (and probably also Windows File Explorer) uses to find out the status.

![image](https://github.com/user-attachments/assets/604f4c10-8ff5-4fed-bc60-68712e95c9fc)

![image](https://github.com/user-attachments/assets/ddab11d6-45f8-42d4-a7dc-5dbd6570842f)

0x5bfa commented on src/Files.App/Utils/Cloud/CloudDrivesDetector.cs:
```diff
@@ -108,6 +108,7 @@ private static Task<IEnumerable<ICloudProvider>> DetectGenericCloudDrive()
 						"Creative Cloud Files" => CloudProviders.AdobeCreativeCloud,
 						"ownCloud" => CloudProviders.ownCloud,
 						"ProtonDrive" => CloudProviders.ProtonDrive,
+						"kDrive" => CloudProviders.kDrive,
```
Don't you have to add this?

```c#
if (driveType.StartsWith("kDrive"))
	driveType = "kDrive";
```

mafra99 replied:
I don't think this is necessary because the ApplicationName in the registry is exactly kDrive
![image](https://github.com/user-attachments/assets/3720c77e-ea06-4893-a5c8-e46f96e48348)

For the others (e.g. iCloudDrive and OneDrive) there is sometimes an ID or name behind it
![image](https://github.com/user-attachments/assets/483a5f2f-57d8-4069-87fa-da388bcf32a6)
![image](https://github.com/user-attachments/assets/2a62cfb6-c439-4cc2-9e63-399372b87140)

0x5bfa replied:
Thanks for letting me know!

yaira2 commented:
@mafra99 does this still work if the user changes the location from kDrive settings?

mafra99 commented:
I have just reinstalled kDrive and now the entry in the registry no longer contains an application name, but only an ID. As a result, kDrive is no longer recognized on my device.
![image](https://github.com/user-attachments/assets/cfec69f3-1846-4029-907c-d68637e48c74)


I can think of two other ways to recognize kDrive anyway: 
1. Using the DefaultIcon under Computer\HKEY_CLASSES_ROOT\CLSID\<>\DefaultIcon to check whether the path contains kDrive
![image](https://github.com/user-attachments/assets/c8276b42-329e-4293-a198-01f8b7199032)

2. Via the entry in Computer\HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\Windows\CurrentVersion\Explorer\SyncRootManager
![image](https://github.com/user-attachments/assets/b7c64c1f-565c-4c05-aadd-1cd5825e1344)

To use the entry from SyncRootManager seems safer to me, I have pushed a commit that contains the changes for this.


In the second commit I cleaned up the code a bit and fixed a few warnings.

The method GetDriveType is new and returns the drive type, e.g. iCloudDrive or kDrive. This should make the method DetectGenericCloudDrive a little simpler.
Compared to before, the DriveType is now returned directly as soon as one is detected and the checks for the other DriveTypes are no longer executed.

mafra99 commented:
> @mafra99 does this still work if the user changes the location from kDrive settings?

@yaira2 thanks, that actually didn't work. should also be fixed

hishitetsu commented on src/Files.App/Utils/Cloud/CloudDrivesDetector.cs:
```diff
@@ -345,5 +329,33 @@ private static Task<IEnumerable<ICloudProvider>> DetectAutodeskDrive()
+			if (driveIdentifier.StartsWith("ProtonDrive"))
+				return "ProtonDrive";
+
+			// Nextcloud specific
+			if (!string.IsNullOrEmpty(appNameFromNamespace) && appNameFromNamespace == "Nextcloud")
+				return appNameFromNamespace;
```
The order of Nextcloud specific and Drive specific are swapped, is it no problem? If it is, L335 should be moved just before here.

mafra99 replied:
The order of the checks should not matter.
But now the registry only has to be used if it cannot be identified by the driveIdentifier.

I have moved L335 👍

hishitetsu reviewed (approved):
LGTM, code wise.

0x5bfa reviewed (approved):
For the record, there's [GetSyncRootInformationForId WinRT method](https://learn.microsoft.com/uwp/api/windows.storage.provider.storageprovidersyncrootmanager.getsyncrootinformationforid) to retrieve sync root manager info but accessing the registry directly should be much faster.

LGTM, code wise.
